pandas/indexes/base.py

```diff
@@ -2886,6 +2888,9 @@ def join(self, other, how='left', level=None, return_indexers=False):
         elif how == 'outer':
             join_index = self.union(other)
 
+        if sort:
```

Comment:
@jreback There is not really an interaction between ``sort`` and `how`. It is just that when `sort=False` (the default) not additional sorting is done, and the order of the result depends on the type of join. When you set ``sort=True``, you explicitly sort. And that regard, I find the current code of 

```
if sort:
    join_index = join_index.sort_values()
```

very clear. 
The only thing is that when ``how='outer'``, you would have sorted twice (as `join_index` is already sorted at this point in that case). But I wouldn't care about that, since it is not the default, and you can just not set `sort` to True to prevent that.